os/fs/smartfs/smartfs_smart.c

```diff
@@ -227,52 +238,57 @@ static int smartfs_open(FAR struct file *filep, const char *relpath, int oflags,
 	 */
 
 	if (ret == OK) {
-		/* The name exists -- but is is a file or a directory ? */
+		/* The name exists -- but is it a file or a directory ? */
 
 		if (sf->entry.flags & SMARTFS_DIRENT_TYPE_DIR) {
 			/* Can't open a dir as a file! */
-
 			ret = -EISDIR;
 			goto errout_with_buffer;
 		}
 
 		/* It would be an error if we are asked to create it exclusively */
-
 		if ((oflags & (O_CREAT | O_EXCL)) == (O_CREAT | O_EXCL)) {
 			/* Already exists -- can't create it exclusively */
 
 			ret = -EEXIST;
 			goto errout_with_buffer;
 		}
 
-		/* The file exists already.
-		 * If the file has been requested to be opened in TRUNCATE mode, then shrink the file.
-		 * However, if APPEND mode has also been requested, do not shrink the file to 0.
+		/* The file exists already */
+
+		/* Test if the file is opened in APPEND mode. If yes, then seek to the end of the file.
+		 * Seeking will also automatically update the header in the sector buffer
 		 */
+		if (oflags & O_APPEND) {
+			smartfs_seek_internal(fs, sf, 0, SEEK_END);
+		}
+
+		/* If the file has been requested to be opened in TRUNCATE mode, then shrink the file.*/
 
-		/* TODO: This method for checking the open modes is not in full accordance with POSIX standard.
+		/* NOTE: This method for checking the open modes is not in full accordance with POSIX standard.
 		 *       POSIX standard truncates the file i.e. shrinks to 0 irrespective of other included flags.
 		 */
-
-		if (((oflags & O_TRUNC) != 0) && ((oflags & O_APPEND) == 0)) {
+		if (oflags & O_TRUNC) {
 			/* Truncate the file to length 0 as part of the open */
 			ret = smartfs_shrinkfile(fs, sf, 0);
 			if (ret < 0) {
 				goto errout_with_buffer;
 			}
 		}
+
 	} else if (ret == -ENOENT) {
 		/* The file does not exist.  Were we asked to create it? */
+
 		if ((oflags & O_CREAT) == 0) {
 			/* No.. then we fail with -ENOENT */
-
```

Comment:
I think it is better to add refactoring changes for comments, etc in a seperate commit..